MatrixSDK/Data/Store/MXMemoryStore/MXMemoryStore.m

```diff
@@ -168,40 +168,47 @@ - (NSString *)partialTextMessageOfRoom:(NSString *)roomId
     return roomStore.partialTextMessage;
 }
 
-- (NSArray<MXReceiptData*> *)getEventReceipts:(NSString*)roomId eventId:(NSString*)eventId sorted:(BOOL)sort
+- (void)loadReceiptsForRoom:(NSString *)roomId completion:(void (^)(void))completion
 {
-    NSMutableArray* receipts = [[NSMutableArray alloc] init];
+    [self getOrCreateReceiptsFor:roomId];
     
-    NSMutableDictionary* receiptsByUserId = receiptsByRoomId[roomId];
-    
-    if (receiptsByUserId)
+    if (completion)
     {
-        @synchronized (receiptsByUserId)
+        dispatch_async(dispatch_get_main_queue(), ^{
+            completion();
+        });
+    }
+}
+
+- (void)getEventReceipts:(NSString *)roomId eventId:(NSString *)eventId sorted:(BOOL)sort completion:(void (^)(NSArray<MXReceiptData *> * _Nonnull))completion
+{
+    [self loadReceiptsForRoom:roomId completion:^{
+        NSDictionary<NSString *, MXReceiptData *> *receiptsByUserId = self->receiptsByRoomId[roomId];
+        
+        if (receiptsByUserId)
         {
-            for (NSString* userId in receiptsByUserId)
+            @synchronized (receiptsByUserId)
             {
-                MXReceiptData* receipt = receiptsByUserId[userId];
-
-                if (receipt && [receipt.eventId isEqualToString:eventId])
+                NSArray<MXReceiptData*> *receipts = [[receiptsByUserId allValues] filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"eventId == %@", eventId]];
```

Comment:
Now that we have a completion block for this method, can we execute this block on a separated thread?
I remembered that operation was expensive.

Reply:
Before doing that, please measure the duration of this call in MatrixHQ.

Reply:
Looks negligible like under 1 ms. Anyway, i've moved that into another queue.